Autoloading, dependency loading and middleware registry cleanup (#1301)
lostisland/faraday: UPGRADING.md

* We've setup an [Awesome Faraday](https://github.com/lostisland/awesome-faraday) repository, where you can find and discover adapters.
  We also highlighted their unique features and level of compliance with Faraday's features.

### Autoloading and dependencies

> Faraday have until now provided and relied on a complex dynamic dependencies system.

Reviewer: ```suggestion
Faraday has until now provided and relied on a complex dynamic dependencies system.
```

This would allow to reference classes and require dependencies only when needed (e.g. when running the first request) based
on the middleware/adapters used.
As part of Faraday v2.0, we've removed all external dependencies, which means we don't really need this anymore.

> This change should not affect you directly, but if you're registering middleware then beware of the new syntax:

Reviewer: ```suggestion
This change should not affect you directly, but if you're registering middleware then be aware of the new syntax:
```

```ruby
# `name` here can be anything you want.
# `klass` is your custom middleware class.

> # This method can be also called on `Faraday::Adapter`, `Faraday::Request` and `Faraday::Response`

Reviewer: ```suggestion
# This method can also be called on `Faraday::Adapter`, `Faraday::Request` and `Faraday::Response`
```
